File: kt/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/DiagnosticBag.kt
package dev.phytolizer.minsk.analysis

import dev.phytolizer.minsk.analysis.syntax.SyntaxKind
import dev.phytolizer.minsk.analysis.text.TextSpan
import kotlin.reflect.KClass

internal class DiagnosticBag {
    private val _diagnostics = mutableListOf<Diagnostic>()
    fun toList(): List<Diagnostic> = _diagnostics

    private fun report(span: TextSpan, message: String) {
        _diagnostics.add(Diagnostic(span, message))
    }

    fun reportBadCharacter(position: Int, char: Char) {
        report(TextSpan(position, 1), "Bad character in input: '$char'")
    }

    fun reportInvalidInt(span: TextSpan, text: String, type: KClass<out Any>) {
        report(span, "The number '$text' doesn't fit in '$type'")
    }

    fun reportUnexpectedToken(span: TextSpan, expected: SyntaxKind, actual: SyntaxKind) {
        report(span, "Expected next token to be $expected, got $actual instead")
    }

    fun reportUndefinedBinaryOperator(
        span: TextSpan,
        op: String,
        leftType: KClass<out Any>,
        rightType: KClass<out Any>,
    ) {
        report(span, "The binary operator '$op' isn't defined for '$leftType' and '$rightType'")
    }

    fun reportUndefinedUnaryOperator(span: TextSpan, op: String, operandType: KClass<out Any>) {
        report(span, "The unary operator '$op' isn't defined for '$operandType'")
    }

    fun addAll(diagnostics: Collection<Diagnostic>) {
        _diagnostics.addAll(diagnostics)
    }

    fun reportUndefinedName(span: TextSpan, name: String) {
        report(span, "Undefined name '$name'")
    }

    fun reportVariableAlreadyDeclared(span: TextSpan, name: String) {
        report(span, "Name '$name' is already declared in the same scope")
    }

    fun reportCannotConvert(span: TextSpan, fromType: KClass<out Any>, toType: KClass<out Any>) {
        report(span, "Cannot convert from '$fromType' to '$toType'")
    }

    fun reportCannotAssign(span: TextSpan, name: String) {
        report(span, "Cannot assign to read-only variable '$name'")
    }
}

File: cs/Minsk/CodeAnalysis/DiagnosticBag.cs
using System.Collections;
using Minsk.CodeAnalysis.Syntax;
using Minsk.CodeAnalysis.Text;

namespace Minsk.CodeAnalysis;

internal sealed class DiagnosticBag : IEnumerable<Diagnostic>
{
    private readonly List<Diagnostic> _diagnostics = new();

    public IEnumerator<Diagnostic> GetEnumerator()
    {
        return _diagnostics.GetEnumerator();
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    private void Report(TextSpan span, string message)
    {
        var diagnostic = new Diagnostic(span, message);
        _diagnostics.Add(diagnostic);
    }

    public void ReportBadCharacter(int position, char c)
    {
        var message = $"Bad character in input: {c}";
        var span = new TextSpan(position, 1);
        Report(span, message);
    }

    public void AddRange(IEnumerable<Diagnostic> diagnostics)
    {
        _diagnostics.AddRange(diagnostics);
    }

    public void ReportInvalidInt(TextSpan span, string text, Type type)
    {
        var message = $"The number {text} doesn't fit in {type}";
        Report(span, message);
    }

    public void ReportUnexpectedToken(TextSpan span, SyntaxKind expectedKind, SyntaxKind actualKind)
    {
        var message = $"Expected next token to be {expectedKind}, got {actualKind} instead";
        Report(span, message);
    }

    public void ReportUndefinedBinaryOperator(TextSpan span, SyntaxToken op, Type leftType, Type rightType)
    {
        var message = $"Binary operator '{op.Text}' is not defined for types '{leftType}' and '{rightType}'";
        Report(span, message);
    }

    public void ReportUndefinedUnaryOperator(TextSpan span, SyntaxToken op, Type operandType)
    {
        var message = $"Unary operator '{op.Text}' is not defined for type '{operandType}'";
        Report(span, message);
    }

    public void ReportUndefinedName(TextSpan span, string name)
    {
        var message = $"Undefined name '{name}'";
        Report(span, message);
    }

    public void ReportVariableAlreadyDeclared(TextSpan span, string name)
    {
        var message = $"Name '{name}' is already declared in this scope";
        Report(span, message);
    }

    public void ReportCannotConvert(TextSpan span, Type fromType, Type toType)
    {
        var message = $"Cannot convert type '{fromType}' to '{toType}'";
        Report(span, message);
    }

    public void ReportCannotAssign(TextSpan span, string name)
    {
        var message = $"Cannot assign to read-only variable '{name}'";
        Report(span, message);
    }
}
